bson/src/main/org/bson/internal/ProvidersCodecRegistry.java

```diff
@@ -50,18 +58,22 @@ public <T> Codec<T> get(final Class<T> clazz, final CodecRegistry registry) {
         return null;
     }
 
-    @SuppressWarnings({"rawtypes" })
+    @SuppressWarnings({"unchecked"})
     public <T> Codec<T> get(final ChildCodecRegistry<T> context) {
-        if (!codecCache.containsKey(context.getCodecClass())) {
+        CodecCacheKey codecCacheKey = new CodecCacheKey(context.getCodecClass(), context.getTypes());
+        if (!codecCache.containsKey(codecCacheKey)) {
             for (CodecProvider provider : codecProviders) {
                 Codec<T> codec = provider.get(context.getCodecClass(), context);
                 if (codec != null) {
-                    return codecCache.putIfMissing(context.getCodecClass(), codec);
+                    if (codec instanceof Parameterizable && context.getTypes() != null) {
+                        codec = (Codec<T>) ((Parameterizable) codec).parameterize(context, context.getTypes());
```

Comment:
Some dicey casting here.  I think it's ok but I wonder if we can do better.

Reply:
Is it an error to have a parameterizable codec with an empty list of types? - Should that scenario be handled at the codec level? Or at least document that parameterize wont be called if the types list is empty.

Reply:
So here we should we just pass context.getTypes() and leave it to the instance to handle. Or we should document that `Parameterizable#parameterize` won't be called with an empty list.

Reply:
In the first commit this was a check for null rather than isEmpty, but was changed as a result of the suggestion to make `ChildCodecRegistry#types` non-nullable.  @stIncMale given @rozza 's comment, should we rethink that suggestion?

Reply:
I think not null was the right choice - its just that a comment is needed in the javadoc for Parameterizable otherwise 
future implementers could get caught out.

Reply:
ok, so javadoc on Parameterizable, plus a `!types.isEmpty()` check for all implementations of the new CodecRegistry method (and Javadoc change there too)?

Reply:
sgtm

Reply:
The `!types.isEmpty()` check is weird: it is usually expected that if a collection is accepted, then an empty collection may be passed. I think it should not be introduced. Also, `RecordCodec` is currently dealing with empty `types` just fine, and itself uses `List.of()` in its `RecordCodec(final Class<T> clazz, final CodecRegistry codecRegistry)` constructor.

Reply:
I guess I can update the Iterable and Map Codec to do the same.

Reply:
If I understand you correctly @stIncMale , we will need to add 

```
        if (types.isEmpty()) {
            return this;
        }
```

at the start of every implementation of `Parameterizable#parameterize`. 

But we should only be calling `Parameterizable#parameterize` as a result of a call to `CodecRegistry#get(java.lang.Class<T>, java.util.List<java.lang.reflect.Type>)`, but without the isEmpty check we end up calling it in as a result of a call to `CodecRegistry#get(java.lang.Class<T>)` as well.

Reply:
We discussed with @jyemin and agreed that:
- `CodecRegistry.get(Class<T> clazz, List<Type> typeArguments)` should check `clazz.getTypeParameters().length == typeArguments.size()` and on top of that check `!typeArguments.isEmpty()`.
- `Parameterizable.parameterize` should check `!typeArguments.isEmpty()`.
- `ChildCodecRegistry.types`/`LazyCodec.types` should become `@Nullable List<Type>` and must not be empty. As a result, `ChildCodecRegistry.getTypes` should return `Optional<List<Type>>`.

Reply:
> CodecRegistry.get(Class<T> clazz, List<Type> typeArguments) should check clazz.getTypeParameters().length == typeArguments.size() and on top of that check !typeArguments.isEmpty().

Also check for null. Done

> Parameterizable.parameterize should check !typeArguments.isEmpty().

I just documented it.  All the implementations already seem to check the size so isEmpty check isn't necessary

> ChildCodecRegistry.types/LazyCodec.types should become @Nullable List<Type> and must not be empty. As a result, ChildCodecRegistry.getTypes should return Optional<List<Type>>.

Done

Reply:
> All the implementations already seem to check the size so isEmpty check isn't necessary

`RecordCodec.parametrize` just passes `types` to the constructor of `RecordCodec`, which does the `types.size() != clazz.getTypeParameters().length` check. However, this check does not prevent `types` from being empty.

A different question: do you think annotating fields and constructor parameters with `@Nullable` (relevant for `ChildCodecRegistry`, `LazyCodec`, `CodecCacheKey`) is not useful?

Reply:
It's just that our `Nullable` annotation is in driver-core, so we don't really use `Nullable` in bson module at this point